overlord/devicestate/devicestate.go

```diff
@@ -1075,3 +1076,43 @@ func CreateRecoverySystem(st *state.State, label string) (*state.Change, error)
 	chg.AddAll(ts)
 	return chg, nil
 }
+
+// InstallFinish creates a change that will finish the install for the given
+// label and volumes. This includes writing missing volume content, seting
+// up the bootloader and installing the kernel.
+func InstallFinish(st *state.State, label string, onVolumes map[string]*client.InstallVolume) (*state.Change, error) {
```

Comment:
the issue of using client.InstallVolume here, is not so much importing client, we do it already, is just that the uncleanliness of the double Structure field here is annoying. I think either in daemon or here we should turn these into normal gadget.Volume? or maybe we should rethink our decision of introducing client.InstallVolume ??

Reply:
Indeed, it's a bit cumbersome. Right now we only do it to amend each `gadget.VolumeStructure` with `{Unencrypted,}Device`. I think we have the options:
1. the smallest change would be to just add `{Unencrypted,}Device` to gadget.VolumeStructure and add a comment that it will never be part of the yaml (reverting the idea to introduce client.InstallVolume)
2. we could add an extra map that amends `map[string]*gadget.Volume` that just contains the device information, something like `map[string]*[]struct{Device,UnencryptedDevice string}` and the list would corresponds to the `[]VolumeStructre` from the `map[string]*gadget.Volume`
3. we could have a `map[string]*struct{Vol gadget.Volume, Device string} `but that would also be the same double struct as before just more explicit so not a win

Not sure if there is more that I forgot? Happy with either of them.

Reply:
At this point I think I'm fine with either 1. or 2.,  about 1. we should also consider that we can give these fields only json tags and not yaml ones (tests might need tweaks though?), we still want clear comments about them though

Reply:
I talked to Alfonso and I think (1) for now is the easiest route so we decided to start there. I updated the code now.